runtime/tchannel_server.go

```diff
@@ -64,197 +83,339 @@ func (ncr netContextRouter) Handle(ctx netContext.Context, call *tchannel.Inboun
-		}
-		s.logger.Error("Unexpected tchannel system error",
-			zap.String("service", service),
-			zap.String("method", method),
-			zap.String("error", err.Error()),
+		c.endpoint.Logger.Error("Could not create arg2reader for inbound request", zap.Error(err))
+		return errors.Wrapf(err, "Could not create arg2reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
 		)
-		return call.Response().SendSystemError(errors.New("Server Error"))
 	}
-
-	if err := EnsureEmpty(treader, "reading request body"); err != nil {
+	c.reqHeaders, err = ReadHeaders(treader)
+	if err != nil {
 		_ = treader.Close()
-
-		return errors.Wrapf(err, "could not ensure arg3reader is empty for inbound call: %s::%s", service, method)
+		c.endpoint.Logger.Error("Could not read headers for inbound request", zap.Error(err))
+		return errors.Wrapf(err, "Could not read headers for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+	}
+	if err := EnsureEmpty(treader, "reading request headers"); err != nil {
+		_ = treader.Close()
+		c.endpoint.Logger.Error("Could not ensure arg2reader is empty for inbound request", zap.Error(err))
+		return errors.Wrapf(err, "Could not ensure arg2reader is empty for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
 	}
 	if err := treader.Close(); err != nil {
-		return errors.Wrapf(err, "could not close arg3reader is empty for inbound call: %s::%s", service, method)
+		c.endpoint.Logger.Error("Could not close arg2reader for inbound request", zap.Error(err))
+		return errors.Wrapf(err, "Could not close arg2reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
 	}
+	return nil
+}
 
-	structWireValue, err := resp.ToWire()
+// readReqBody reads request body from arg3
+func (c *tchannelInboundCall) readReqBody() (wireValue wire.Value, err error) {
+	treader, err := c.call.Arg3Reader()
 	if err != nil {
-		// If we could not write the body then we should do something else
-		// instead.
+		c.endpoint.Logger.Error("Could not create arg3reader for inbound request", zap.Error(err))
+		err = errors.Wrapf(err, "Could not create arg3reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+		return
+	}
+	buf := GetBuffer()
+	defer PutBuffer(buf)
+	if _, err = buf.ReadFrom(treader); err != nil {
+		_ = treader.Close()
+		c.endpoint.Logger.Error("Could not read from arg3reader for inbound request", zap.Error(err))
+		err = errors.Wrapf(err, "Could not read from arg3reader for inbound %s.%s (%s) request",
+			c.endpoint.EndpointID, c.endpoint.HandlerID, c.endpoint.Method,
+		)
+		return
+	}
+	wireValue, err = protocol.Binary.Decode(bytes.NewReader(buf.Bytes()), wire.TStruct)
+	if err != nil {
+		c.endpoint.Logger.Error("Could not decode arg3 for inbound request", zap.Error(err))
```

Comment:
there are so many `Error()` here that need to be downgraded to `Warn()`

Rule of thumb : 
 - If we are dealing with network input or any input foreign to this process then it's a `Warn()`
 - If we are dealing with the application developer passing the wrong arguments or doing the wrong thing ( aka its a bug in my own application ) then it's an `Error()` 
 - If we are dealing with writing out of the process ; aka writes to the TCP socket that can fail then again its something foreign to the process or application developer and its a `Warn()`

Reply:
To me it makes sense that reading or writing from/to the network should be an application error (It's the server's fault it couldn't read or write to the network).  If the wrong thing (i.e. missing headers, wrong body fields/format, etc.) was sent by the foreign caller then that's a client error or application warning.